File: small_problems/medium2/triangle_sides.rb
# Triangle Sides
# A triangle can be classified as follows:
# - Equilateral: All 3 sides are equal length
# - Isosceles: 2 sides are equal length, the 3rd is different
# - Scalene: All 3 sides are different lengths
# Rules for a valid triangle:
# - The two shortest sides combined must be longer than the longest side
# - All sides must be longer than 0

# Write a method that takes 3 integers representing the lengths of the triangle
#   and returns a symbol:
#   - :equilateral, :isosceles, :scalene, or :invalid

# Input: 3 integers
# - Represent the lengths of the 3 sides of the 'triangle'
# Output: A symbol 
# - Represents the type of triangle (or invalid if triangle is invalid)

# Examples:
# triangle(3, 3, 3) == :equilateral
# triangle(3, 3, 1.5) == :isosceles
# triangle(3, 4, 5) == :scalene
# triangle(0, 3, 3) == :invalid
# triangle(3, 1, 1) == :invalid

# Data:
# - The 3 integers given as input each representing a side
# - First calculate if triangle is valid
# - If valid, then calculate which type it is
# - Maybe store the lengths in an array

# Algorithm:
# Given 3 integers as input, length1..length3:
# Calculate whether the triangle is valid.
#   All lengths must be greater than 0, AND
#     the sum of the two smallest lengths must be greater than the largest length
# If triangle is invalid, return :invalid. Otherwise:
# Calculate the type of triangle.
#   If all lengths have the same value -> return equilateral
#   If all lengths have different values -> return scalene
#   * If 2 lengths are identical and 1 length is different -> return isosceles

# Compress lengths to only include unique values:
#   - If size == 1 -> equilateral
#   - If size == 2 -> isosceles
#   - If size == 3 -> scalene

def triangle(length1, length2, length3)
  lengths = [length1, length2, length3]
  
  return :invalid unless valid_triangle?(lengths)
  
  case lengths.uniq.size
  when 1 then :equilateral
  when 2 then :isosceles
  when 3 then :scalene
  end
end

def valid_triangle?(lengths)
  lengths.size == 3 && lengths.all?(&:positive?) && 
    lengths.min(2).sum > lengths.max
end


p triangle(3, 3, 3) == :equilateral
p triangle(3, 3, 1.5) == :isosceles
p triangle(3, 4, 5) == :scalene
p triangle(0, 3, 3) == :invalid
p triangle(3, 1, 1) == :invalid

File: javaScriptExercises/medium/triangleSides.js
// Triangle Sides
// A triangle is classifed as follows:
// 1) Equilateral - All 3 sides are equal
// 2) Isosceles   - 2 sides are equal, 1 is different
// 3) Scalene     - All 3 sides are different

// For a triangle to be valid, the 2 shortest side lengths must be greater than the
// length of the longest side, and every side must be longer than 0.

// Write a function that takes 3 numbers representing the side lengths of a 
// triangle and returns a string classifying the triangle.
// > equilateral, isosceles, scalene, or invalid.

// Rules/Patterns:
// - A triangle is only considered valid if:
//   1) Its 2 shortest sides are cumulatively longer than its longest side.
//      eg. [2, 3, 7] => 2 + 3 (5) is less than 7 => invalid
//   2) All sides are positive (greater than 0)
// - If the triangle is valid, classify it based on the uniqueness of its lengths:
//   - 3 unique => scalene
//   - 2 unique => isosceles
//   - 1 unique => equilateral

// Data: 
// Input: 3 numbers 
// 1) Validate
// => Array of numbers eg. [3, 4, 5]
//    a) Validate that lengths are all > 0
//    b) Validate that the two shortest sides > the longest side
// 2) Classify
// => Convert the array of lengths to a Set (stores primitive values uniquely by
//    value)
//    - Classify based on the #size of the Set (3 => scalene, 2 => isosceles,
//      1 => equilateral)
// Output: String (classifying the triangle type)

// Algorithm:
// Given 3 numbers as input:
// 1) Determine the validity of the triangle:
//    a) Place the numbers in an array
//    b) Interrogate - All numbers must be > 0
//    c) Sort the array in ascending order
//    d) Confirm that array (first 2 elements) > array (last element)
//    - If at any point validation fails, return 'invalid'

// 2) If the triangle is valid, classify it:
//    a) Convert the lengths array to a Set
//    b) Match the Set#size to an integer between 1-3:
//    => 1 => equilateral
//    => 2 => isosceles
//    => 3 => scalene

function triangle(lengthA, lengthB, lengthC) {
  let lengths = [ lengthA, lengthB, lengthC ];

  if (!isTriangleValid(lengths)) return 'invalid';

  let uniqueLengths = new Set(lengths);
  switch (uniqueLengths.size) {
    case 1:
      return 'equilateral';
    case 2:
      return 'isosceles';
    case 3:
      return 'scalene';
  }
}

function isTriangleValid(lengths) {
  lengths = lengths.toSorted((a, b) => Number(a) - Number(b));

  return lengths.length === 3 && 
         lengths.every(length => length > 0) &&
         lengths[0] + lengths[1] > lengths[2];
}

// Examples:
console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"